benchmarks/WolfSheep/wolf_sheep.py

```diff
@@ -94,26 +93,42 @@ class GrassPatch(CellAgent):
     A patch of grass that grows at a fixed rate and it is eaten by sheep
     """
 
-    def __init__(self, unique_id, model, fully_grown, countdown):
+    @property
+    def fully_grown(self):
+        return self._fully_grown
+
+    @fully_grown.setter
+    def fully_grown(self, value: bool) -> None:
+        self._fully_grown = value
+
+        if not value:
+            self.model.simulator.schedule_event_relative(
+                self.set_fully_grown, self.grass_regrowth_time
+            )
+
+    def __init__(self, unique_id, model, fully_grown, countdown, grass_regrowth_time):
         """
+        TODO:: fully grown can just be an int --> so one less param (i.e. countdown)
+
         Creates a new patch of grass
 
         Args:
             fully_grown: (boolean) Whether the patch of grass is fully grown or not
             countdown: Time for the patch of grass to be fully grown again
+            grass_regrowth_time : time to fully regrow grass
+            countdown : Time for the patch of grass to be fully regrown if fully grown is False
         """
         super().__init__(unique_id, model)
-        self.fully_grown = fully_grown
-        self.countdown = countdown
+        self._fully_grown = fully_grown
+        self.grass_regrowth_time = grass_regrowth_time
 
-    def step(self):
         if not self.fully_grown:
-            if self.countdown <= 0:
-                # Set as fully grown
-                self.fully_grown = True
-                self.countdown = self.model.grass_regrowth_time
-            else:
-                self.countdown -= 1
+            self.model.simulator.schedule_event_relative(
```

Comment:
If the grow time is longer than the tick time, won't this schedule multiple events before it's fully grown?

Reply:
I guess I have no idea how this works.

- The step method is removed.
- One event is scheduled in the init
- There are properties and setters
- there is a set_fully_grown() method

I don't see where after the init (which happens only on model initialisation right?) grass is regrown when it's being eaten. Shouldn't that be a method that can be triggered when the grass is eaten?

Reply:
Oh wait does grass die if it's eaten? And will a new grass be created when not?

I was under the presumption that grass exists on each cell, is either grown or now, and covers the full space. But it's actually sparse, and some grass is created on initialisation, and spreads to other cells, creating new Agents on those locations. When it's eaten it's removed. Just like the sheep if they are eaten by a wolf.

If this is right never mind my earlier comments, this can be marked resolved.

Reply:
I have simplified the code to just

```python
    @fully_grown.setter
    def fully_grown(self, value: bool) -> None:
        self._fully_grown = value

        if not value:
            self.model.simulator.schedule_event_relative(
                setattr,
                self.grass_regrowth_time,
                function_args=[self, "fully_grown", True],
            )
```

This is used in conjunction with a `Sheep`, who when eating grass does `grass_patch.fully_grown = False`. So, this triggers the setter. The setter, in turn, schedules the event for when it becomes fully grown again. This is just a `setattr` call with the provided arguments. Because value in this case is True, we don't schedule a new event. 

Hope this helps.

Reply:
Right, looks good. Thanks!